File: src/main/java/io/github/chindeaone/collectiontracker/tracker/collection/LeaderboardManager.java
package io.github.chindeaone.collectiontracker.tracker.collection;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.categories.overlay.LeaderboardConfig;
import io.github.chindeaone.collectiontracker.commands.CollectionTracker;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class LeaderboardManager {
    private static volatile List<LeaderboardEntry> currentLeaderboard = List.of();
    private static final Map<String, List<LeaderboardEntry>> skillLeaderboards = new ConcurrentHashMap<>();

    public static void set(List<LeaderboardEntry> entries) {
        currentLeaderboard = entries;
    }

    public static void setSkillLeaderboard(String skill, List<LeaderboardEntry> entries) {
        if (skill == null || entries == null) return;
        skillLeaderboards.put(skill.toLowerCase(), entries);
    }

    private static LeaderboardEntry getPlayerEntryRaw(long amount) {
        List<LeaderboardEntry> lb = currentLeaderboard;
        if (lb.isEmpty()) {
            return null;
        }

        int index = findBinaryIndex(lb, amount);
        if (index < lb.size()) {
            return lb.get(index);
        }

        return null;
    }

    public static LeaderboardEntry getPlayerEntry(long amount) {
        return getPlayerEntryRaw(amount);
    }

    public static LeaderboardEntry getPlayerEntry(String skill, long amount) {
        List<LeaderboardEntry> lb = skillLeaderboards.getOrDefault(skill.toLowerCase(), List.of());
        if (lb.isEmpty()) return null;

        int index = findBinaryIndex(lb, amount);
        return index < lb.size() ? lb.get(index) : null;
    }

    private static LeaderboardEntry getNextRankEntryRaw(long amount) {
        List<LeaderboardEntry> lb = currentLeaderboard;
        if (lb.isEmpty()) {
            return null;
        }

        int index = findBinaryIndex(lb, amount);

        if (index > 0) {
            return lb.get(index - 1);
        }

        return null;
    }

    private static LeaderboardEntry getNextRankEntryRaw(String skill, long amount) {
        List<LeaderboardEntry> lb = skillLeaderboards.getOrDefault(skill.toLowerCase(), List.of());
        if (lb.isEmpty()) {
            return null;
        }

        int index = findBinaryIndex(lb, amount);


        if (index > 0) {
            return lb.get(index - 1);
        }

        return null;
    }

    public static LeaderboardEntry getNextRankEntry(long amount) {
        List<LeaderboardEntry> lb = currentLeaderboard;
        if (lb.isEmpty()) {
            return null;
        }

        // Custom goal
        if (ConfigAccess.isCustomGoalEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            LeaderboardConfig.CustomGoalEntry goalEntry = ConfigAccess.getCustomGoalEntry("gemstone");

            if (goalEntry != null) {
                LeaderboardConfig.CustomGoalType goalType = ConfigAccess.getCustomGoalType();

                if (goalType == LeaderboardConfig.CustomGoalType.POSITION && goalEntry.position != null) {
                    LeaderboardEntry playerEntry = getPlayerEntryRaw(amount);
                    // default if player already passed the goal position
                    if (playerEntry != null && playerEntry.rank() < goalEntry.position) {
                        return getNextRankEntryRaw(amount);
                    }
                    return getEntryAtPosition(goalEntry.position);
                } else if (goalType == LeaderboardConfig.CustomGoalType.AMOUNT && goalEntry.amount != null) {
                    LeaderboardEntry playerEntry = getPlayerEntryRaw(amount);
                    // default if player's collection is greater than the custom goal amount
                    if (playerEntry != null && amount > goalEntry.amount) {
                        return getNextRankEntryRaw(amount);
                    }
                    LeaderboardEntry entry = getNextRankEntryRaw(goalEntry.amount);
                    // If goal amount exceeds all players, return custom goal entry
                    if (entry == null && !lb.isEmpty()) {
                        return new LeaderboardEntry("Custom Goal", 0, goalEntry.amount);
                    }
                    return entry;
                }
            }
        }

        // Default
        return getNextRankEntryRaw(amount);
    }

    public static LeaderboardEntry getNextRankEntry(String skill, long amount) {
        List<LeaderboardEntry> lb = skillLeaderboards.getOrDefault(skill.toLowerCase(), List.of());
        if (lb.isEmpty()) return null;

        // Custom goal
        if (ConfigAccess.isCustomGoalEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            LeaderboardConfig.CustomGoalEntry goalEntry = ConfigAccess.getCustomGoalEntry(skill);

            if (goalEntry != null) {
                LeaderboardConfig.CustomGoalType goalType = ConfigAccess.getCustomGoalType();

                if (goalType == LeaderboardConfig.CustomGoalType.POSITION && goalEntry.position != null) {
                    LeaderboardEntry playerEntry = getPlayerEntry(skill, amount);
                    // default if player already passed the goal position
                    if (playerEntry != null && playerEntry.rank() < goalEntry.position) {
                        return getNextRankEntryRaw(skill, amount);
                    }
                    return getSkillEntryAtPosition(skill, goalEntry.position);
                } else if (goalType == LeaderboardConfig.CustomGoalType.AMOUNT && goalEntry.amount != null) {
                    LeaderboardEntry playerEntry = getPlayerEntry(skill, amount);
                    // default if player's skill xp is greater than the custom goal amount
                    if (playerEntry != null && amount > goalEntry.amount) {
                        return getNextRankEntryRaw(skill, amount);
                    }
                    LeaderboardEntry entry = getNextRankEntryRaw(skill, goalEntry.amount);
                    // If goal amount exceeds all players, return custom goal entry
                    if (entry == null && !lb.isEmpty()) {
                        return new LeaderboardEntry("Custom Goal", 0, goalEntry.amount);
                    }
                    return entry;
                }
            }
        }

        // Default
        return getNextRankEntryRaw(skill, amount);
    }

    public static LeaderboardEntry getPlayerEntry() {
        return getPlayerEntryRaw(TrackingRates.collectionAmount);
    }

    public static LeaderboardEntry getNextRankEntry() {
        // Custom goal
        if (ConfigAccess.isCustomGoalEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            LeaderboardConfig.CustomGoalEntry goalEntry = ConfigAccess.getCustomGoalEntry(CollectionTracker.collection);

            if (goalEntry != null) {
                LeaderboardConfig.CustomGoalType goalType = ConfigAccess.getCustomGoalType();

                if (goalType == LeaderboardConfig.CustomGoalType.POSITION && goalEntry.position != null) {
                    LeaderboardEntry playerEntry = getPlayerEntryRaw(TrackingRates.collectionAmount);
                    // default if player already passed the goal position
                    if (playerEntry != null && playerEntry.rank() < goalEntry.position) {
                        return getNextRankEntryRaw(TrackingRates.collectionAmount);
                    }
                    return getEntryAtPosition(goalEntry.position);
                } else if (goalType == LeaderboardConfig.CustomGoalType.AMOUNT && goalEntry.amount != null) {
                    LeaderboardEntry playerEntry = getPlayerEntryRaw(TrackingRates.collectionAmount);
                    // default if player's collection is greater than the custom goal amount
                    if (playerEntry != null && TrackingRates.collectionAmount > goalEntry.amount) {
                        return getNextRankEntryRaw(TrackingRates.collectionAmount);
                    }
                    LeaderboardEntry entry = getNextRankEntryRaw(goalEntry.amount);
                    // If goal amount exceeds all players, return custom goal entry
                    if (entry == null && !currentLeaderboard.isEmpty()) {
                        return new LeaderboardEntry("Custom Goal", 0, goalEntry.amount);
                    }
                    return entry;
                }
            }
        }

        // Default
        return getNextRankEntryRaw(TrackingRates.collectionAmount);
    }

    public static LeaderboardEntry getEntryAtPosition(int position) {
        List<LeaderboardEntry> lb = currentLeaderboard;
        if (lb.isEmpty() || position < 1 || position > lb.size()) {
            return null;
        }
        return lb.get(position - 1);
    }

    public static LeaderboardEntry getNextRankEntryForSkill(String skill, long skillXp) {
        return getNextRankEntry(skill, skillXp);
    }

    public static LeaderboardEntry getSkillEntryAtPosition(String skill, int position) {
        List<LeaderboardEntry> lb = skillLeaderboards.getOrDefault(skill.toLowerCase(), List.of());
        if (lb.isEmpty() || position < 1 || position > lb.size()) {
            return null;
        }
        return lb.get(position - 1);
    }

    public static int findBinaryIndex(List<LeaderboardEntry> lb, long targetAmount) {
        int index = Collections.binarySearch(lb, new LeaderboardEntry("", 0, targetAmount),
                (a, b) -> Long.compare(b.amount(), a.amount()));

        if (index < 0) {
            index = -index - 1;
        }
        return index;
    }

    public static void clear() {
        currentLeaderboard = List.of();
        skillLeaderboards.clear();
    }

    public static boolean isEmpty() {
        return currentLeaderboard.isEmpty();
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/tracker/collection/LeaderboardManager.kt
package io.github.chindeaone.collectiontracker.tracker.collection

import io.github.chindeaone.collectiontracker.config.ConfigAccess
import io.github.chindeaone.collectiontracker.commands.CollectionTracker

object LeaderboardManager {
    @Volatile
    private var currentLeaderboard = mutableListOf<LeaderboardEntry>()
    @Volatile
    private var skillLeaderboards = mutableMapOf<String, List<LeaderboardEntry>>()

    @JvmStatic
    fun set(entries: MutableList<LeaderboardEntry>) {
        currentLeaderboard = entries
    }

    @JvmStatic
    fun setSkillLeaderboard(skill: String, entries: MutableList<LeaderboardEntry>) {
        skillLeaderboards[skill.lowercase()] = entries
    }

    private fun getPlayerEntryRaw(amount: Long): LeaderboardEntry? {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return null
        }

        val index = findBinaryIndex(lb, amount)
        if (index < lb.size) {
            return lb[index]
        }

        return null
    }

    fun getPlayerRank(amount: Long): Int {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return -1
        }

        return findBinaryIndex(lb, amount) + 1
    }

    fun getPlayerEntry(skill: String, amount: Long): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) return null

        val index = findBinaryIndex(lb, amount)
        return if (index < lb.size) lb[index] else null
    }

    @JvmStatic
    fun getPlayerRank(skill: String, amount: Long): Int {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) {
            return -1
        }

        return findBinaryIndex(lb, amount) + 1
    }

    private fun getNextRankEntryRaw(amount: Long): LeaderboardEntry? {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return null
        }

        val index = findBinaryIndex(lb, amount)
        if (index > 0) {
            return lb[index - 1]
        }

        return null
    }

    private fun getPreviousRankEntryRaw(amount: Long): LeaderboardEntry? {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return null
        }

        val index = findBinaryIndex(lb, amount)
        if (index < lb.size - 1) {
            return lb[index + 1]
        }

        return null
    }

    private fun getNextRankEntryRaw(skill: String, amount: Long): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) {
            return null
        }

        val index = findBinaryIndex(lb, amount)
        if (index > 0) {
            return lb[index - 1]
        }

        return null
    }

    private fun  getPreviousRankEntryRaw(skill: String, amount: Long): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) {
            return null
        }

        val index = findBinaryIndex(lb, amount)
        if (index < lb.size - 1) {
            return lb[index + 1]
        }

        return null
    }

    fun getNextRankEntry(amount: Long): LeaderboardEntry? {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return null
        }

        // Custom position
        if (ConfigAccess.isCustomPositionEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            val position = ConfigAccess.getCustomPositionEntry("gemstone")

            if (position != null) {
                val playerEntry = getPlayerEntryRaw(amount)
                // default if player already passed the goal position
                if (playerEntry != null && playerEntry.rank < position) {
                    return getNextRankEntryRaw(amount)
                }
                return getEntryAtPosition(position)
            }
        }

        // Default
        return getNextRankEntryRaw(amount)
    }

    fun getNextRankEntry(skill: String, amount: Long): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) {
            return null
        }

        // Custom position
        if (ConfigAccess.isCustomPositionEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            val position = ConfigAccess.getCustomPositionEntry(skill)

            if (position != null) {
                val playerEntry = getPlayerEntry(skill, amount)
                // default if player already passed the goal position
                if (playerEntry != null && playerEntry.rank < position) {
                    return getNextRankEntryRaw(skill, amount)
                }
                return getSkillEntryAtPosition(skill, position)
            }
        }

        // Default
        return getNextRankEntryRaw(skill, amount)
    }

    fun getPreviousRankEntry(skill: String, amount: Long): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty()) {
            return null
        }

        return getPreviousRankEntryRaw(skill, amount)
    }

    fun getPreviousRankEntry(amount: Long): LeaderboardEntry? {
        val lb = currentLeaderboard
        if (lb.isEmpty()) {
            return null
        }

        return getPreviousRankEntryRaw(amount)
    }

    @JvmStatic
    fun getPlayerRank(): Int {
        return getPlayerRank(TrackingRates.collectionAmount)
    }

    @JvmStatic
    fun getNextRankEntry(): LeaderboardEntry? {
        // Custom position
        if (ConfigAccess.isCustomPositionEnabled() && !ConfigAccess.getCustomGoals().isEmpty()) {
            val position = ConfigAccess.getCustomPositionEntry(CollectionTracker.collection)

            if (position != null) {
                val playerEntry = getPlayerEntryRaw(TrackingRates.collectionAmount)
                // default if player already passed the goal position
                if (playerEntry != null && playerEntry.rank < position) {
                    return getNextRankEntryRaw(TrackingRates.collectionAmount)
                }
                return getEntryAtPosition(position)
            }
        }

        // Default
        return getNextRankEntryRaw(TrackingRates.collectionAmount)
    }

    @JvmStatic
    fun getPreviousRankEntry(): LeaderboardEntry? {
        return getPreviousRankEntryRaw(TrackingRates.collectionAmount)
    }

    fun getEntryAtPosition(position: Int): LeaderboardEntry? {
        val lb = currentLeaderboard

        if (lb.isEmpty() || position < 1 || position > lb.size) {
            return null
        }
        return lb[position - 1]
    }

    @JvmStatic
    fun getNextRankEntryForSkill(skill: String, skillXp: Long): LeaderboardEntry? {
        return getNextRankEntry(skill, skillXp)
    }

    @JvmStatic
    fun getPreviousRankEntryForSkill(skill: String, skillXp: Long): LeaderboardEntry? {
        return getPreviousRankEntry(skill, skillXp)
    }

    fun getSkillEntryAtPosition(skill: String, position: Int): LeaderboardEntry? {
        val lb = skillLeaderboards.getOrDefault(skill.lowercase(), listOf())
        if (lb.isEmpty() || position < 1 || position > lb.size) {
            return null
        }
        return lb[position - 1]
    }

    fun findBinaryIndex(lb: List<LeaderboardEntry>, targetAmount: Long): Int {
        var index = lb.binarySearch(
            LeaderboardEntry("", 0, targetAmount, ConfigAccess.isIncludeWipedProfilesEnabled()),
            { a, b -> b.amount.compareTo(a.amount) })

        if (index < 0) {
            index = -index - 1
        }
        return index
    }

    @JvmStatic
    fun clear() {
        currentLeaderboard = mutableListOf()
        skillLeaderboards.clear()
    }

    @JvmStatic
    fun isEmpty(): Boolean {
        return currentLeaderboard.isEmpty()
    }
}

data class LeaderboardEntry(var username: String, var rank: Int, var amount: Long, var wiped: Boolean)